### ISSUE TITLE
vertical interpolation of tm5 data is wrong

### ISSUE THREAD

**Issue by etiennesky** [EC-Earth/ece2cmor3#603]
Hi @goord @tommibergman @treerink when trying to cmorize co2 from tm5 I found a bug in  interpolate_plev().

The problem is that Ngl.vinth2p() expects plevo in mb . When cmorizing to plev19 the values are given in Pa rather than in mb - however it seems that plev39 is ok.

I was going to suggest to divide the argument to Ngl.vinth2p()  by 100, but this will not work when cmorizing to plev39.  Not sure which is the proper solution here since the pressure level data request is in Pa not mb.

I think we should fix the values of plev39_ to have it in Pa (as defined in the data request), then fix the call to  Ngl.vinth2p() 

```
# Pressure levels
plev19_ = numpy.array([100000.,92500.,85000.,70000.,60000.,50000.,40000.,30000.,25000.,20000.,15000.,10000.,7000.,5000.,3000.,2000.,1000.,500.,100.])
plev39_ = numpy.array([1000.,925.,850.,700.,600.,500.,400.,300.,250.,200.,170.,150.,130.,115.,100.,90.,80.,70.,50.,30.,20.,15.,10.,7.,5.,3.,2.,1.5,1.,0.7,0.5,0.4,0.3,0.2,0.15,0.1,0.07,0.05,0.03])
```

**Comment by etiennesky** [comment]
@aearamos

**Comment by goord** [comment]
Absolutely, all pressures are to be in Pa.

**Comment by plesager** [comment]
Related to #590, which also involves vertical interpolation to plev19.

**Comment by etiennesky** [comment]
indeed this is the same problem as in #590, since I also have negative values

original file:
```
cdo infon ../18500101/fc0/runtime/output/tm5/001/co2_AERmon_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001-185012.nc 
    -1 :       Date     Time   Level Gridsize    Miss :     Minimum        Mean     Maximum : Parameter name
     1 : 1850-01-16 12:00:00       1    10800       0 :  0.00027060  0.00028430  0.00030765 : co2           
     2 : 1850-01-16 12:00:00       2    10800       0 :  0.00027607  0.00028427  0.00029902 : co2           
     3 : 1850-01-16 12:00:00       3    10800       0 :  0.00028117  0.00028413  0.00029097 : co2           
     4 : 1850-01-16 12:00:00       4    10800       0 :  0.00028225  0.00028408  0.00028747 : co2           
     5 : 1850-01-16 12:00:00       5    10800       0 :  0.00028236  0.00028404  0.00028556 : co2           
     6 : 1850-01-16 12:00:00       6    10800       0 :  0.00028244  0.00028399  0.00028464 : co2           
     7 : 1850-01-16 12:00:00       7    10800       0 :  0.00028260  0.00028396  0.00028453 : co2           
     8 : 1850-01-16 12:00:00       8    10800       0 :  0.00028375  0.00028399  0.00028406 : co2           
     9 : 1850-01-16 12:00:00       9    10800       0 :  0.00028400  0.00028400  0.00028404 : co2           
    10 : 1850-01-16 12:00:00      10    10800       0 :  0.00028399  0.00028400  0.00028400 : co2           
```
incorrectly cmorized:
```
cdo infon ../18500101/fc0/runtime/cmor_outputs/cmor_1/CMIP/EC-Earth-Consortium/EC-Earth3-CC/piControl/r1i1p1f1/Amon/co2/gn/v20200317/co2_Amon_EC-Earth3-CC_piControl_r1i1p1f1_gn_185001-185001.nc.1
    -1 :       Date     Time   Level Gridsize    Miss :     Minimum        Mean     Maximum : Parameter name
     1 : 1850-01-16 12:00:00  100000    10800       0 :   -0.065022  0.00062394    0.095744 : co2           
     2 : 1850-01-16 12:00:00   92500    10800       0 :   -0.060080  0.00059824    0.088526 : co2           
     3 : 1850-01-16 12:00:00   85000    10800       0 :   -0.055139  0.00057253    0.081309 : co2           
     4 : 1850-01-16 12:00:00   70000    10800       0 :   -0.045256  0.00052113    0.066874 : co2           
     5 : 1850-01-16 12:00:00   60000    10800       0 :   -0.038668  0.00048686    0.057251 : co2           
     6 : 1850-01-16 12:00:00   50000    10800       0 :   -0.032080  0.00045258    0.047628 : co2           
     7 : 1850-01-16 12:00:00   40000    10800       0 :   -0.025491  0.00041831    0.038005 : co2           
     8 : 1850-01-16 12:00:00   30000    10800       0 :   -0.018903  0.00038404    0.028382 : co2           
     9 : 1850-01-16 12:00:00   25000    10800       0 :   -0.015609  0.00036691    0.023570 : co2           
    10 : 1850-01-16 12:00:00   20000    10800       0 :   -0.012314  0.00034977    0.018759 : co2           
    11 : 1850-01-16 12:00:00   15000    10800       0 :  -0.0090202  0.00033263    0.013947 : co2           
    12 : 1850-01-16 12:00:00   10000    10800       0 :  -0.0057260  0.00031550   0.0091356 : co2           
    13 : 1850-01-16 12:00:00    7000    10800       0 :  -0.0037495  0.00030522   0.0062487 : co2           
    14 : 1850-01-16 12:00:00    5000    10800       0 :  -0.0024318  0.00029836   0.0043240 : co2           
    15 : 1850-01-16 12:00:00    3000    10800       0 :  -0.0011141  0.00029151   0.0023994 : co2           
    16 : 1850-01-16 12:00:00    2000    10800       0 : -0.00045525  0.00028808   0.0014371 : co2           
    17 : 1850-01-16 12:00:00    1000    10800       0 :  0.00020359  0.00028465  0.00047478 : co2           
    18 : 1850-01-16 12:00:00     500    10800       0 :  0.00028236  0.00028404  0.00028563 : co2           
    19 : 1850-01-16 12:00:00     100    10800       0 :  0.00028297  0.00028397  0.00028438 : co2           
```

correctly cmorized (after my fix):
```
cdo infon ../18500101/fc0/runtime/cmor_outputs/cmor_1/CMIP/EC-Earth-Consortium/EC-Earth3-CC/piControl/r1i1p1f1/Amon/co2/gn/v20200317/co2_Amon_EC-Earth3-CC_piControl_r1i1p1f1_gn_185001-185001.nc
    -1 :       Date     Time   Level Gridsize    Miss :     Minimum        Mean     Maximum : Parameter name
     1 : 1850-01-16 12:00:00  100000    10800       0 :  0.00020359  0.00028465  0.00047478 : co2           
     2 : 1850-01-16 12:00:00   92500    10800       0 :  0.00024568  0.00028438  0.00040695 : co2           
     3 : 1850-01-16 12:00:00   85000    10800       0 :  0.00027465  0.00028421  0.00034291 : co2           
     4 : 1850-01-16 12:00:00   70000    10800       0 :  0.00028228  0.00028408  0.00029155 : co2           
     5 : 1850-01-16 12:00:00   60000    10800       0 :  0.00028234  0.00028406  0.00028628 : co2           
     6 : 1850-01-16 12:00:00   50000    10800       0 :  0.00028236  0.00028404  0.00028563 : co2           
     7 : 1850-01-16 12:00:00   40000    10800       0 :  0.00028242  0.00028402  0.00028501 : co2           
     8 : 1850-01-16 12:00:00   30000    10800       0 :  0.00028244  0.00028400  0.00028466 : co2           
     9 : 1850-01-16 12:00:00   25000    10800       0 :  0.00028246  0.00028398  0.00028462 : co2           
    10 : 1850-01-16 12:00:00   20000    10800       0 :  0.00028251  0.00028397  0.00028457 : co2           
    11 : 1850-01-16 12:00:00   15000    10800       0 :  0.00028257  0.00028396  0.00028454 : co2           
    12 : 1850-01-16 12:00:00   10000    10800       0 :  0.00028297  0.00028397  0.00028438 : co2           
    13 : 1850-01-16 12:00:00    7000    10800       0 :  0.00028350  0.00028398  0.00028416 : co2           
    14 : 1850-01-16 12:00:00    5000    10800       0 :  0.00028378  0.00028399  0.00028406 : co2           
    15 : 1850-01-16 12:00:00    3000    10800       0 :  0.00028394  0.00028400  0.00028402 : co2           
    16 : 1850-01-16 12:00:00    2000    10800       0 :  0.00028400  0.00028400  0.00028403 : co2           
    17 : 1850-01-16 12:00:00    1000    10800       0 :  0.00028400  0.00028400  0.00028401 : co2           
    18 : 1850-01-16 12:00:00     500    10800       0 :  0.00028400  0.00028400  0.00028400 : co2           
    19 : 1850-01-16 12:00:00     100    10800       0 :  0.00028399  0.00028400  0.00028400 : co2           
```

**Comment by plesager** [comment]
> I think we should fix the values of plev39_ to have it in Pa (as defined in the data request), then fix the call to Ngl.vinth2p()

Yes, that's the fix, because it is used to defined the zaxis that ends up in the netcdf file. 
Ideally `plev39` should be retrieved from the data request, but I'm not sure that's possible.

@treerink this probably needs to be in the next release. I'm going to test it along my tests for #590

**Comment by etiennesky** [comment]
fixing this in the add-co2+co2mass (but not touching the plev39_ variable)

**Comment by plesager** [comment]
How then?

**Comment by etiennesky** [comment]
```
diff --git a/ece2cmor3/tm52cmor.py b/ece2cmor3/tm52cmor.py
index 7124116..210667f 100644
--- a/ece2cmor3/tm52cmor.py
+++ b/ece2cmor3/tm52cmor.py
@@ -556,7 +556,8 @@ def interpolate_plev(pressure_levels,dataset,psdata,varname):
     data = data[:,::-1,:,:]
 
     interpolation=1 #1 linear, 2 log, 3 loglog
-    interpolated_data = Ngl.vinth2p(data,hyam,hybm,pressure_levels,psdata[:,:,:],interpolation,p0mb,1,True)
+    # divide pressure_levels by 100 to get in mb
+    interpolated_data = Ngl.vinth2p(data,hyam,hybm,pressure_levels/100,psdata[:,:,:],interpolation,p0mb,1,True)
     return interpolated_data
```

**Comment by etiennesky** [comment]
this i in the add-co2+co2mass - works for plev19.

But you should probably test with a real request for plev39 and check everything is ok.

**Comment by etiennesky** [comment]
Although I guess the fix could be to change the values of plev19 - probably this fix will mess up plev39 interpolation.

**Comment by plesager** [comment]
No, your fix is correct but partial.  `plev39` declaration needs to be converted to Pa.

**Comment by plesager** [comment]
I just realized that this variables are default variables and used only if the `TABLE_coordinate.json` file is not found. No warning if this is no the case. Ouch.

**Comment by plesager** [comment]
In #590, the interpolation is done on the correct levels since `plev39` is found in `ece2cmor3/ece2cmor3/resources/cmip6-cmor-tables/Tables/CMIP6_coordinate.json`. Still, it would be nice to:
- convert `plev39_` default value to Pa,
- add a warning when the default values are used, although I wonder if the logic is flawed. Is it wise to fall back on default value? Shouldn't it be an error if the `coordinate.json` file is not found?

A commit in one of my branches has it all: 766fa05

**Comment by etiennesky** [comment]
but I guess you also need to divide the plev by 100 also for the interpolation to be done correctly?

**Comment by treerink** [comment]
I will include the @plesager changes 766fa05 into the `add-co2+co2mass` branch. So I can test all in one.

**Comment by plesager** [comment]
> but I guess you also need to divide the plev by 100 also for the interpolation to be done correctly?

Yes, it goes on top of your change.

**Comment by treerink** [comment]
In 139671d8d98a71b8331bf65e4657109253ef2b6d the 766fa05 changes have been added. So everything in this issue is now covered in the `add-co2+co2mass` branch connected to #589. Therefore closing this issue.

**Comment by plesager** [comment]
I doubt the vertical interpolation is correct. If I look at the CH4, here is what I get for one longitude. Same data, on model levels on the right and on plev19 on the left:

![ch4-ML-and-plev19](https://user-images.githubusercontent.com/11032052/77157032-e0f9cb80-6aa0-11ea-9091-d1bbf520e585.png)

How can we explain going from [1e-8, 1e-6] on model levels to [-1e-3, 1e-5] on plev19. There are negatives values.

**Comment by plesager** [comment]
And here when I use the same color scale as ML for the plev19 plot:

![ch4-plev19-other-scale](https://user-images.githubusercontent.com/11032052/77157707-2e2a6d00-6aa2-11ea-97c7-a632f28feb89.png)

**Comment by plesager** [comment]
This is a puzzling. How can CO2 be correct and not CH4?

**Comment by plesager** [comment]
I used the corrected EC-Earth (sp) and ece2cmor3 (corrected as above).

**Comment by etiennesky** [comment]
out of curiosity, is the Pa variable ok? THe fix I put in EC-Earth was for 1hr output, and might have had a bad impact for you.

**Comment by plesager** [comment]
sp is fine:

![ch4-sp](https://user-images.githubusercontent.com/11032052/77166701-08f32a00-6ab5-11ea-885b-90897ead441d.png)

**Comment by etiennesky** [comment]
I have uploaded the tm5 output and cmorized co2 and co2mass, could you test it on your system?

ftp://bscesftp.bsc.es/etourign/ec-earth/co2/

I you want to send me your files I can test as well.

**Comment by plesager** [comment]
FTP: Access refused and I cannot share my data easily.  Anyway a simple zonal average _on model level_ (bottom plot) looks fine (and will look good if converted to pressure afterwards):

![simple-zonal-ch4](https://user-images.githubusercontent.com/11032052/77180706-9a20cb80-6aca-11ea-85df-f448420226c5.png)

Doing the conversion to pressure along vertical interpolation first (as wrapped in ece2cmor3) creates the problem. I wonder if the large drop in surface pressure above greenland and antartica are not handled correctly. @noije does it make any sense to you?

**Comment by etiennesky** [comment]
with a web browser and ncftp I am able to access the files on my ftp server...

**Comment by noije** [comment]
@plesager Yes, I think the problem is related to the treatment of the points with high surface elevation. See my comment in #590.

**Comment by etiennesky** [comment]
I was able to cmorize ch4 without problems using my installation, will try with a fresh copy of the add-co2+co2mass .

![tm5-ch4](https://user-images.githubusercontent.com/1018556/77326481-28d75780-6d1a-11ea-9d13-4b4ee91e2dc9.png)

On the left is raw output (34 levels), on the right is cmorized output (19 plevels)

**Comment by etiennesky** [comment]
Fine for me with current trunk.  

```
cns60490@login0: /gpfs/scratch/cns60/cns60490/a2jk/LOG_a2jk > cdo infon ../18500101/fc4/runtime/cmor_outputs/cmor_1/CMIP/EC-Earth-Consortium/EC-Earth3-CC/piControl/r5i1p1f1/Amon/ch4/gn/v20200317/ch4_Amon_EC-Earth3-CC_piControl_r5i1p1f1_gn_185001-185001.nc 
    -1 :       Date     Time   Level Gridsize    Miss :     Minimum        Mean     Maximum : Parameter name
     1 : 1850-01-16 12:00:00  100000    10800       0 :  8.5488e-07  9.7284e-07  2.5504e-06 : ch4           
     2 : 1850-01-16 12:00:00   92500    10800       0 :  8.4723e-07  9.6867e-07  2.2304e-06 : ch4           
     3 : 1850-01-16 12:00:00   85000    10800       0 :  8.5111e-07  9.7967e-07  1.9105e-06 : ch4           
     4 : 1850-01-16 12:00:00   70000    10800       0 :  8.8042e-07  1.0324e-06  1.3029e-06 : ch4           
     5 : 1850-01-16 12:00:00   60000    10800       0 :  9.1086e-07  1.0889e-06  1.3560e-06 : ch4           
     6 : 1850-01-16 12:00:00   50000    10800       0 :  9.7436e-07  1.1534e-06  1.4515e-06 : ch4           
     7 : 1850-01-16 12:00:00   40000    10800       0 :  1.0282e-06  1.1974e-06  1.5018e-06 : ch4           
     8 : 1850-01-16 12:00:00   30000    10800       0 :  1.0424e-06  1.3464e-06  1.7158e-06 : ch4           
     9 : 1850-01-16 12:00:00   25000    10800       0 :  1.0514e-06  1.5821e-06  2.1890e-06 : ch4           
    10 : 1850-01-16 12:00:00   20000    10800       0 :  1.0497e-06  1.9303e-06  2.4827e-06 : ch4           
    11 : 1850-01-16 12:00:00   15000    10800       0 :  1.0566e-06  2.0343e-06  2.6382e-06 : ch4           
    12 : 1850-01-16 12:00:00   10000    10800       0 :  8.2168e-07  1.4199e-06  2.9166e-06 : ch4           
    13 : 1850-01-16 12:00:00    7000    10800       0 :  7.1268e-07  1.2492e-06  2.7863e-06 : ch4           
    14 : 1850-01-16 12:00:00    5000    10800       0 :  6.0524e-07  7.5000e-07  1.0962e-06 : ch4           
    15 : 1850-01-16 12:00:00    3000    10800       0 :  5.4966e-07  6.9590e-07  8.9571e-07 : ch4           
    16 : 1850-01-16 12:00:00    2000    10800       0 :  5.4706e-07  6.7532e-07  8.4676e-07 : ch4           
    17 : 1850-01-16 12:00:00    1000    10800       0 :  6.5977e-07  8.6308e-07  1.1131e-06 : ch4           
    18 : 1850-01-16 12:00:00     500    10800       0 :  4.8727e-07  7.4617e-07  1.0884e-06 : ch4           
    19 : 1850-01-16 12:00:00     100    10800       0 :  1.6878e-07  2.9846e-07  4.4552e-07 : ch4           
```
If I omit the fix to divide the plevels by 100 I get negative values in the lower levels, upper levels are ok (just as before with co2)

```
cns60490@login0: /gpfs/scratch/cns60/cns60490/a2jk/LOG_a2jk > cdo infon ../18500101/fc4/runtime/cmor_outputs/cmor_1/CMIP/EC-Earth-Consortium/EC-Earth3-CC/piControl/r5i1p1f1/Amon/ch4/gn/v20200317/ch4_Amon_EC-Earth3-CC_piControl_r5i1p1f1_gn_185001-185001.nc 
    -1 :       Date     Time   Level Gridsize    Miss :     Minimum        Mean     Maximum : Parameter name
     1 : 1850-01-16 12:00:00  100000    10800       0 : -6.5404e-05  1.7759e-05  0.00096705 : ch4           
     2 : 1850-01-16 12:00:00   92500    10800       0 : -6.0382e-05  1.6487e-05  0.00089389 : ch4           
     3 : 1850-01-16 12:00:00   85000    10800       0 : -5.5361e-05  1.5216e-05  0.00082072 : ch4           
     4 : 1850-01-16 12:00:00   70000    10800       0 : -4.5318e-05  1.2672e-05  0.00067438 : ch4           
     5 : 1850-01-16 12:00:00   60000    10800       0 : -3.8622e-05  1.0977e-05  0.00057682 : ch4           
     6 : 1850-01-16 12:00:00   50000    10800       0 : -3.1927e-05  9.2812e-06  0.00047926 : ch4           
     7 : 1850-01-16 12:00:00   40000    10800       0 : -2.5231e-05  7.5856e-06  0.00038170 : ch4           
     8 : 1850-01-16 12:00:00   30000    10800       0 : -1.8536e-05  5.8900e-06  0.00028414 : ch4           
     9 : 1850-01-16 12:00:00   25000    10800       0 : -1.5188e-05  5.0422e-06  0.00023536 : ch4           
    10 : 1850-01-16 12:00:00   20000    10800       0 : -1.1840e-05  4.1944e-06  0.00018658 : ch4           
    11 : 1850-01-16 12:00:00   15000    10800       0 : -8.4923e-06  3.3466e-06  0.00013780 : ch4           
    12 : 1850-01-16 12:00:00   10000    10800       0 : -5.1445e-06  2.4988e-06  8.9023e-05 : ch4           
    13 : 1850-01-16 12:00:00    7000    10800       0 : -3.1359e-06  1.9902e-06  5.9756e-05 : ch4           
    14 : 1850-01-16 12:00:00    5000    10800       0 : -1.7968e-06  1.6511e-06  4.0244e-05 : ch4           
    15 : 1850-01-16 12:00:00    3000    10800       0 : -4.5768e-07  1.3119e-06  2.0732e-05 : ch4           
    16 : 1850-01-16 12:00:00    2000    10800       0 :  2.1187e-07  1.1424e-06  1.0976e-05 : ch4           
    17 : 1850-01-16 12:00:00    1000    10800       0 :  8.5488e-07  9.7284e-07  2.5504e-06 : ch4           
    18 : 1850-01-16 12:00:00     500    10800       0 :  9.7436e-07  1.1534e-06  1.4515e-06 : ch4           
    19 : 1850-01-16 12:00:00     100    10800       0 :  8.2168e-07  1.4199e-06  2.9166e-06 : ch4           
cdo infon: Processed 205200 values from 1 variable over 1 timestep ( 0.03s )
```

**Comment by plesager** [comment]
How is sp in your 3D field?

**Comment by etiennesky** [comment]
```
 cdo infon ../../../../../../../../output/tm5/001/ps_AERmon_EC-Earth3-AerChem_a2jk_r1i1p1f1_gn_185001-185012.nc 
    -1 :       Date     Time   Level Gridsize    Miss :     Minimum        Mean     Maximum : Parameter name
     1 : 1850-01-16 12:00:00       0    10800       0 :      52771.      96595.  1.0295e+05 : ps            
cdo infon: Processed 10800 values from 1 variable over 1 timestep ( 0.01s )
```

around 67000 over Antartica and Greenland and 10000 over ocean

![tm5-ps](https://user-images.githubusercontent.com/1018556/77336143-65f61680-6d27-11ea-8bdb-3f006ea90da9.png)

**Comment by etiennesky** [comment]
I can upload it all to my ftp server if you like

**Comment by plesager** [comment]
Our data look quite similar, and we are using the same cmorization software. I cannot see where is the difference.

**Comment by tommibergman** [comment]
Hi, I am back from the sick leave. I can have a look as well. Are you working with the trunk or is there a branch?

**Comment by plesager** [comment]
The master has everything needed. There is no branch where further work is being tested.

**Comment by treerink** [comment]
Yes for this use the latest master

**Comment by plesager** [comment]
I rerun everything. I got better results (see also #590 for O3), the data range of the interpolated results is realistic now (although the max value on 19 plev is double than the original data). Left: model levels, center: vertical interpolation on 19 pressure levels, right: zonal mean on 39 pressure levels:

![CH4 from 2020-03-24 master](https://user-images.githubusercontent.com/11032052/77428783-f476a080-6dd8-11ea-8836-9c2aa9251add.png)

Like O3 (#590, small negative at the surface), there is still some artifact with CH4, visible in the 19 levels plot if we use the same color scale (model levels on the left, plev19 on the right):

![CH4 artifact 2020-03-24 from master](https://user-images.githubusercontent.com/11032052/77430385-64862600-6ddb-11ea-80ae-66bb3adb2fdc.png)

**Comment by noije** [comment]
At what latitude does the artifact appear? Could it be associated with erroneous treatment of mountains in the zonal mean?

**Comment by plesager** [comment]
Indeed: 
![surface CH4 from 2020-03-24 master](https://user-images.githubusercontent.com/11032052/77438486-2b06e800-6de6-11ea-931d-26aeddd2a922.png)

**Comment by plesager** [comment]
The fix I tried before: 0f2ca78b0a1

**Comment by plesager** [comment]
The last attempt: https://github.com/EC-Earth/ece2cmor3/issues/590#issuecomment-601628054

**Comment by plesager** [comment]
Test data (output from TM5) can be found here:

> ec:/nm6/ch4-o3-tm5-test-output.tar

**Comment by tommibergman** [comment]
Could you add the corresponding `ps` also?

**Comment by tommibergman** [comment]
First the zonal means, on the left we have in pressure levels and on the right it is on model levels. The range is not exactly same but very close. The pressure level version shows the Antarctic missing values. 
<img width="1453" alt="Screenshot 2020-03-27 at 16 34 29" src="https://user-images.githubusercontent.com/737412/77767053-29892a00-7049-11ea-8d9e-18a2eda2fa5f.png">

On the left is the ch4 3d on 100000 Pa pressure level. Most of the land masses have been left out, not sure if it should be like that or not. They mostly have pressures which are lower than 100000Pa.  As can be seen on the right the pressures are below 100000Pa, as very little bright red is seen over land areas.

<img width="300" alt="Screenshot 2020-03-27 at 16 36 04" src="https://user-images.githubusercontent.com/737412/77768207-d1532780-704a-11ea-9724-665d802503cd.png"><img width="300" alt="Screenshot 2020-03-27 at 16 42 11" src="https://user-images.githubusercontent.com/737412/77768233-d912cc00-704a-11ea-8006-8477fbed2b80.png">

It would seem to work, in the sense that the plev are strict limits. Any comments?

**Comment by etiennesky** [comment]
It might be worth trying this over a full year (or a few years), since ps can change over time?

**Comment by plesager** [comment]
Seems alright. Isn't the red band above the NaN (white) area in the upper left corner an artifact of ncview? Like when computing contours?

**Comment by tommibergman** [comment]
> Seems alright. Isn't the red band above the NaN (white) area in the upper left corner an artifact of ncview? Like when computing contours?

Yes, that's right, it is just an artefact. I forgot to change that on ncview.

**Comment by etiennesky** [comment]
Look pretty reasonable, tried it for co2 and getting similar patterns in the missing data.  The last level to have missing data is 60000.

Now is this cmor compliant???

**Comment by plesager** [comment]
@tommibergman, have you had a chance to look at the vertically interpolated 3D data, like the 19-plevel CH4 at lon=292.5?

**Comment by tommibergman** [comment]
Well most of the land surfaces are outside the pressure of 100000Pa. And the high point in the Andes is likewise masked. There is high point on the highest pressure is amazon, but the concentrations in the model and pressure level outputs are almost the same in this area. Also there are some other high concentrations, but they are similar to model level output. Figures have equal ranges. There is no indication of anomalous concentrations in the Andes at lower pressure levels  either.

<img width="1217" alt="Screenshot 2020-03-30 at 14 10 52" src="https://user-images.githubusercontent.com/737412/77910483-de188b00-7297-11ea-97ff-7a0ed17e3d6a.png">

**Comment by plesager** [comment]
Ok I see. Thanks. I think we can get your commit into the master and close this issue and #590 as well. @treerink can you merge?

**Comment by tommibergman** [comment]
I made a pull request

**Comment by etiennesky** [comment]
Are you guys sure that setting the values where ps<plevel to NaN is the way to go?

I think it makes more sense to set them to missval instead, so you can do computations with cdo. 

I am able to set them to missval with `cdo setctomiss,nan co2.nc  tmp1.nc`

More importantly, I not sure that NaN is correct for CMOR data? thoughts @goord  ?

**Comment by plesager** [comment]
Internally, to compute the zonal mean, the NaN is convenient. In the final output, I do not know. Should they be (or are they already) set to "missing value" or to "fill value"? CMOR protocol should have an answer.

**Comment by goord** [comment]
Hi Philippe, there shouldn't ne NaN's in the output, they have to be replaced with the FillValue, 1e+40 I believe for CMIP6

**Comment by tommibergman** [comment]
Didn't really think of that, I'll remove the pull request and modify to use the missing values.

**Comment by tommibergman** [comment]
Actually the missing value is 1e20, which can be retrieved from the tables. I fixed the retrieval of this value and made a mask for nans. Now the variable missing data is showing 1e20.  It is already pushed here.

**Comment by treerink** [comment]
Solved by merging #618.
